### ISSUE TITLE
Some jobs failing due to ActiveRecord::Deadlocked when trying to create a ScheduledExecution

### ISSUE THREAD

**Issue by andbar** [rails/solid_queue#162]
We are seeing some failed jobs due to hitting a deadlock when solid queue is trying to create the ScheduledExecution for the job. This is usually happening for us on jobs that are having to be retried due to a throttling constraint we are dealing with from an external api. Here is one example, with part of the backtrace. The job attempts to execute, gets the throttling constraint so it tries to schedule a retry, and it looks like it's trying to do the `ScheduledExecution.create_or_find_by!` on line 40 of `app/models/solid_queue/job/schedulable.rb` when it hits the deadlock on the insert.

<img width="1382" alt="Screenshot 2024-03-01 at 12 19 12 PM" src="https://github.com/basecamp/solid_queue/assets/14093131/79492e6c-466f-4317-bd80-54fccaaa4edf">

Backtrace:
```
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.5/lib/mysql2/client.rb:151:in `_query'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.5/lib/mysql2/client.rb:151:in `block in query'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.5/lib/mysql2/client.rb:150:in `handle_interrupt'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/mysql2-0.5.5/lib/mysql2/client.rb:150:in `query'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/mysql2/database_statements.rb:100:in `block (2 levels) in raw_execute'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1028:in `block in with_raw_connection'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1000:in `with_raw_connection'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/mysql2/database_statements.rb:98:in `block in raw_execute'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:1143:in `log'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/mysql2/database_statements.rb:97:in `raw_execute'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:233:in `execute_and_free'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/mysql2/database_statements.rb:23:in `internal_exec_query'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:153:in `exec_insert'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:191:in `insert'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/query_cache.rb:25:in `insert'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/persistence.rb:588:in `_insert_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/persistence.rb:1252:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/counter_cache.rb:187:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/locking/optimistic.rb:84:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/encryption/encryptable_record.rb:184:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/attribute_methods/dirty.rb:240:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/callbacks.rb:445:in `block in _create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:110:in `run_callbacks'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:952:in `_run_create_callbacks'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/callbacks.rb:445:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/timestamp.rb:114:in `_create_record'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/persistence.rb:1221:in `create_or_update'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/callbacks.rb:441:in `block in create_or_update'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:110:in `run_callbacks'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/callbacks.rb:952:in `_run_save_callbacks'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/callbacks.rb:441:in `create_or_update'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/timestamp.rb:125:in `create_or_update'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/persistence.rb:751:in `save!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/validations.rb:55:in `save!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/transactions.rb:313:in `block in save!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/transactions.rb:365:in `block in with_transaction_returning_status'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:342:in `transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/transactions.rb:361:in `with_transaction_returning_status'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/transactions.rb:313:in `save!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/suppressor.rb:56:in `save!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/persistence.rb:55:in `create!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:918:in `_create!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:118:in `block in create!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:929:in `_scoping'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:467:in `scoping'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:118:in `create!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:229:in `block in create_or_find_by!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:535:in `block in within_new_transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activesupport-7.1.3.2/lib/active_support/concurrency/null_lock.rb:9:in `synchronize'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/transaction.rb:532:in `within_new_transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/connection_adapters/abstract/database_statements.rb:344:in `transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/transactions.rb:212:in `transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation/delegation.rb:105:in `transaction'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/relation.rb:229:in `create_or_find_by!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/activerecord-7.1.3.2/lib/active_record/querying.rb:23:in `create_or_find_by!'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/solid_queue-0.2.1/app/models/solid_queue/job/schedulable.rb:40:in `schedule'
/var/www/shield/vendor/bundle/ruby/3.2.0/gems/solid_queue-0.2.1/app/models/solid_queue/job/executable.rb:65:in `prepare_for_execution'
```

**Comment by rosa** [comment]
@andbar, could you share the `LATEST DETECTED DEADLOCK` section if you run `SHOW ENGINE INNODB STATUS` in the DB server where you have Solid Queue tables?

**Comment by andbar** [comment]
Here you go. Thanks for looking into it. Let me know if you need any more info.

[latest_detected_deadlock.txt](https://github.com/basecamp/solid_queue/files/14465372/latest_detected_deadlock.txt)

**Comment by rosa** [comment]
@andbar, I've looked into this and I see why the deadlock is happening but it's not clear to me why the transaction (2) is locking 361 records in the `index_solid_queue_dispatch_all` index of `solid_queue_scheduled_executions`. Could you let me know: 
- What version of Solid Queue you're running
- What version of MySQL you're running
- Your dispatcher configuration

Thank you! 🙏

**Comment by andbar** [comment]
Yep, here they are:

solid_queue (0.2.1)
8.0.mysql_aurora.3.03.1
```
dispatchers:
    - polling_interval: 1
      batch_size: 500
```

**Comment by rosa** [comment]
Agh, @andbar, I was completely stumped because I thought the code you were running included this change https://github.com/basecamp/solid_queue/pull/155/commits/a30c2cb47a8f872c89f1643f80dec87e5f001691 that I got about 3 weeks ago (GitHub shows last week because I had rebased) and that we're running in production because we're running the branch with support for recurring tasks. I noticed that one while I was working on recurring jobs because I hit a super similar deadlock, and fixed it there. Then, when you reported this one and I looked at your `SHOW ENGINE INNODB STATUS` I was like "how can this possible?!" 😆  In my head, I had already shipped that fix ages ago. I think the deadlock is due to that missing `job_id` in that query. 

I'm going to try to ship a new version with support for recurring jobs and that fix so you can test it. I'm on call this week and had a busy Monday, but hopefully will get to it tomorrow!

**Comment by andbar** [comment]
Ah! Haha, so easy to do. We'll be glad to test that fix with our situation. I'll watch for a new version. Thank you!

**Comment by rosa** [comment]
Thank you so much! I just shipped version 0.2.2 with [this fix](https://github.com/basecamp/solid_queue/pull/164), as I didn't have time to wrap up the recurring jobs PR, so I decided to just extract that fix. Could you try this one and see if you still encounter the deadlock? 🙏  Thank you!

**Comment by andbar** [comment]
Hi, @rosa, we deployed that new version and thought it might have fixed it but unfortunately we got some more deadlocks today. Here's the latest deadlock log from the db, hopefully it helps pinpoint what might be causing it.

[latest_deadlock.txt](https://github.com/basecamp/solid_queue/files/14513488/latest_deadlock.txt)

**Comment by rosa** [comment]
Ouch 😞 Sorry about that, and thanks for the new debug info. I'll continue looking into it.

**Comment by andbar** [comment]
Thanks, @rosa. I haven't had a chance to look at it much myself before this, due to some other needed work in our project that's using this gem, but I'm hoping to be able to dig into it further. I don't have much experience with deadlocks, though, so I'm trying to brush up on that first 😬.

From a first glance, it appears that maybe the issue is locking due to both transactions (the insert and the delete) needing to lock this index: index_solid_queue_dispatch_all? Or more accurately, the insert appears to lock the PRIMARY index first and then tries to acquire a lock on the index_solid_queue_dispatch_all index, while the delete is going in the opposite direction and locks the index_solid_queue_dispatch_all index first and then tries to acquire a lock on the PRIMARY index. Does that sound right?

**Comment by rosa** [comment]
Hey @andbar, sorry for the delay! I haven't forgotten about this one, but I've been working on other stuff in the last few weeks. I'm back looking at this, and I wonder if https://github.com/rails/solid_queue/pull/199 might help in your case. Would you mind trying that one out if you're still using Solid Queue and experiencing this problem?

**Comment by andbar** [comment]
Hi @rosa, I just realized I hadn't responded to you yet, I apologize for that. We ended up moving away from Solid Queue to using something else that was just a better fit for our particular needs due to the nature of the jobs we're running, so unfortunately I won't be able to test that fix. I'm sorry!

**Comment by rosa** [comment]
Hi @andbar! Oh, no worries at all! Thanks so much for letting me know, really appreciate you taking the time to test and report this and your patience through the troubleshooting 🤗

**Comment by paulhtrott** [comment]
Hi @rosa 👋🏽

We tried your fix #199 at [Pressable](https://my.pressable.com), it fixed our deadlock issue

Rails 7.1.2
ruby 3.2.2
MariaDB 10.6.17 (trilogy)

Thank you! 🚀

**Comment by paulhtrott** [comment]
Hi again @rosa. I spoke too soon. #199 did not completely get rid of our deadlock issue, but it did reduce it.

**Comment by rosa** [comment]
Hey @paulhtrott, ohhhh, thanks for letting me know! It's a bummer it didn't solve it completely, but I'll take the reduction. I'll ship that and will start using it in HEY, as we're going to increase the number of scheduled jobs per day in ~4M tomorrow, so hopefully we'll hit this ourselves and that will give me more info to think of other solutions.

**Comment by paulhtrott** [comment]
Hi @rosa we have had zero luck being able to diagnose our deadlock issue. This is how our architecture is structured:

- Docker Container deployed to 4 instances
- Single MariaDB instance (trilogy gem)

Our errors do not show much details outside of, plus a stack trace (attached):

- ActiveRecord::Deadlocked: Trilogy::ProtocolError: 1213: Deadlock found when trying to get lock; try restarting transaction
- solid_queue:start : ActiveRecord::Deadlocked: Trilogy::ProtocolError: 1213: Deadlock found when trying to get lock; try restarting transaction

We have resorted to setting wait_untils for most of our jobs, it seems the delay helps on most occasions, but it is inconvenient in some cases.  

Are there any other details that might be helpful for you? 

[stack_trace.txt](https://github.com/rails/solid_queue/files/15489825/stack_trace.txt)

**Comment by rosa** [comment]
Ohhh, @paulhtrott, that's super helpful! This looks like a different kind of deadlock than the one I tried to fix! Let me try to put together another fix for this one.

**Comment by rosa** [comment]
That deadlock is the same one as https://github.com/rails/solid_queue/pull/229, going to tweak that one a bit and ship.

**Comment by rosa** [comment]
@paulhtrott, I just released version 0.3.3 with a possible fix for this one deadlock: #240.

**Comment by paulhtrott** [comment]
> @paulhtrott, I just released version 0.3.3 with a possible fix for this one deadlock: #240. 

Thank you @rosa , we will give that a try today, I'll report back after a couple day 🎉

**Comment by paulhtrott** [comment]
Hi @rosa I'm back sooner than I wanted to be 😆 . We are still having the issue after 0.3.3. Same two stack traces basically.

**Comment by rosa** [comment]
Ohh, bummer 😞 I'll continue working on it.

**Comment by rosa** [comment]
Oh, @paulhtrott, I realised of something... The query done now to delete records from `solid_queue_ready_executions` should be using the primary key instead of `job_id`, so I wonder if the locks that both transactions are waiting are still in the primary key index or another index... If you have the output from the `LATEST DETECTED DEADLOCK` section when you run `SHOW ENGINE INNODB STATUS`, would you mind sharing it?

Thank you so much again 🙏

**Comment by paulhtrott** [comment]
Hey @rosa! Sure, here is the output

[deadlock.txt](https://github.com/user-attachments/files/15808585/deadlock.txt)

**Comment by paulhtrott** [comment]
Hi @rosa! Just want to see if you have had a chance to look into the deadlock issue? Here is a new deadlock file to show that this is still happening on production.

**Our Setup**
Linux Docker Container -> Four instances 
Puma version: 6.4.3
MariaDB: 10.6.17 -> Single Instance ( solid_queue tables are also here )
Ruby: 3.3.5
Rails: 7.1.3.4
Trilogy: 2.8.1
SolidQueue: 1.0.0.beta

[solid_queue-deadlock.txt](https://github.com/user-attachments/files/17139392/solid_queue-deadlock.txt)

**Comment by iJackUA** [comment]
I've noticed that we are receiving Deadlock error occasionally, only in places where jobs were enqueued in the loops
(solid_queue 1.0.0, Rails 7.2)

smth like this 

```ruby
      items.each do |item|
        #....
        ItemProcessingJob.perform_later(item, #...other params)
      end
```

In terms of best practices, it is clearly not the best code.
Now rewriting it to utilize `ActiveJob.perform_all_later`, let's see would it help or not.

**Comment by iJackUA** [comment]
This issue searching for a workaround  (which at least reduces Deadlocks for MariaDB in Galera Cluster):

Reduce transaction level isolation.
Setting `transaction_isolation="READ-COMMITTED"` seems to resolve deadlocks in my case. 
Default level before that was `REPEATABLE READ`.
But it's up to you to decide is it an appropriate level of tx isolation for your app.

**Comment by rosa** [comment]
Ah, good point! If you're running Solid Queue in its own DB, `READ-COMMITTED` should be completely fine for Solid Queue operation.

**Comment by paulhtrott** [comment]
Yeah, unfortunately that doesn't work in our environment.

**Comment by rosa** [comment]
Closing this one after a long life! To summarise where things stand, since this thread became the home for several different deadlocks:

- The original one, retries creating a `ScheduledExecution` against the dispatcher's `DELETE`, was fixed in three steps by 0.3.3: #164, #199 (the dispatcher deletes scheduled executions by primary key) and #240 (claims delete ready executions by primary key too).
- The ones reported later, inserts waiting on gaps locked by the workers' `SKIP LOCKED` scans, come from InnoDB's gap locking under `REPEATABLE READ`. If Solid Queue has its own database, running that connection with `READ COMMITTED` avoids them; that's now documented in the [README](https://github.com/rails/solid_queue?tab=readme-ov-file#database-configuration) with the `database.yml` snippet. Fewer polling threads and a longer `polling_interval` also help a lot.
- The one deadlock still under investigation, between dispatchers deleting batches of scheduled executions, is tracked in #609.

Nobody has reported here on a 1.x release, so I'm going to close it. If you're still hitting deadlocks on a current version with `READ COMMITTED`, please open a new issue with a fresh `SHOW ENGINE INNODB STATUS` and your configuration.
